cephfs: set metadata on subvolume snapshot

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This helps Monitoring solutions without access to Kubernetes clusters to display the details of thenapshot-name/snapshot-namespace/snapshotcontent-name in their dashboard.

## Is there anything that requires special attention ##
~~Needs: https://github.com/ceph/go-ceph/pull/698~~

~~Just for the CI to pass I have included go-ceph custom branch and hence DNM, but the reviews can continue for other commits.~~



---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
cephfs: add set/Remove subvolume snapshot metadata utility functions

Add utility functions to set/Remove
snapshot-name/snapshot-namespace/snapshotcontent-name metadata on
subvolume snapshots.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
cephfs: set metadata on the subvolume snapshot on create

Set snapshot-name/snapshot-namespace/snapshotcontent-name details
on subvolume snapshots as metadata on create.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
deploy: add --extra-create-metadata arg to csi-snapshotter sidecar

This argument in csi-snapshotter sidecar allows us to receive
snapshot-name/snapshot-namespace/snapshotcontent-name metadata in the
CreateSnapshot() request.

For ex:

csi.storage.k8s.io/volumesnapshot/name
csi.storage.k8s.io/volumesnapshot/namespace
csi.storage.k8s.io/volumesnapshotcontent/name

This is a useful information which can be used depend on the use case we
have at our driver. The features like adding metadata to snapshot image
can consume this based on the need.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: add test case for snapshot metadata validation

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -850,6 +850,77 @@ var _ = Describe(cephfsType, func() {
+				if err != nil {
+					e2elog.Failf("failed to list subvolume snapshots metadata: %v", err)
+				}
+				if metadata[0].volSnapNameKey != snap.Name {
```
check metadata len is not zero before accessing it.

Madhu-1 replied:
same for snapshot and subvolumes list also.

Madhu-1 commented on e2e/cephfs_helper.go:
```diff
@@ -212,6 +212,64 @@ func listCephFSSubVolumes(f *framework.Framework, filesystem, groupname string)
+	f *framework.Framework,
+	filesystem,
+	subvolume,
+	snapname,
```
re-use this helper function to get metadata for both subvolume and snapshot.

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -701,7 +712,7 @@ func (cs *ControllerServer) CreateSnapshot(
 		}
 	}()
-	snap, err := doSnapshot(ctx, parentVolOptions, sID.FsSnapshotName)
+	snap, err := doSnapshot(ctx, req, parentVolOptions, sID.FsSnapshotName)
```
pass only `k8s.GetSnapshotMetadata(parameters)` for from here. we dont need to pass whole `req`

Madhu-1 commented on internal/cephfs/core/snapshot.go:
```diff
@@ -51,6 +51,12 @@ type SnapshotClient interface {
 	// CloneSnapshot clones the snapshot of the subvolume.
 	CloneSnapshot(ctx context.Context, cloneVolOptions *SubVolume) error
+	// SetAllSnapshotMetadata set all the metadata from arg parameters on
+	// Subvolume snapshot.
```
```suggestion
       // subvolume snapshot.
```

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+package core
```
The license header is missing.

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+
+// ListSnapshotMetadata lists custom metadata (key-value pairs) set on the subvolume
+// snapshot in a volume.
+func (s *snapshotClient) ListSnapshotMetadata() (map[string]string, error) {
```
As we dont require list, please remove it.

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+	for k, v := range parameters {
+		err := s.SetSnapshotMetadata(k, v)
+		if err != nil {
+			return fmt.Errorf("failed to set metadata key %q, value %q on subvolume snapshot %v: %w", k, v, s, err)
```
returning `s` in fmt.Errof will return the struct pointer address, not the snapshot details.

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+func (s *snapshotClient) UnsetAllSnapshotMetadata(keys []string) error {
+	for _, key := range keys {
+		err := s.RemoveSnapshotMetadata(key)
+		// TODO: replace string comparison with errno.
```
time to address `TODO` we have this too for all the metadata operations

pkalever replied:
This must be coming from Ceph, Ceph commands are initiated from go-ceph to Ceph and go-ceph actually unmarshal the json output passed from Ceph. So where should be searching for a matching error? That is something I don't know :-(
I'm happy for the adventure but need some lead here.

Maybe @phlogistonjohn @nixpanic might know.

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+		err := s.RemoveSnapshotMetadata(key)
+		// TODO: replace string comparison with errno.
+		if err != nil && !strings.Contains(err.Error(), "No such file or directory") {
+			return fmt.Errorf("failed to unset metadata key %q on subvolume snapshot %v: %w", key, s, err)
```
same comment like above setMetadata menthod

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// SetAllSnapshotMetadata set all the metadata from arg parameters on
+// Subvolume snapshot.
```
`subvolume snapshot`

Madhu-1 reviewed (changes requested):
requesting for changes till we get go-ceph release.

Commit:
cephfs: update subvolume snapshot metadata if snapshot already exists.

Make sure to set metadata when subvolume snapshot exist, i.e. if the
provisioner pod is restarted while createSnapShot is in progress, say it
created the subvolume snapshot but didn't yet set the metadata.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: CephFS validate restore and clone metadata

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
@Mergifyio rebase

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -850,6 +850,220 @@ var _ = Describe(cephfsType, func() {
 				}
 			})
 
+			By("Test subvolume snapshot and restored PVC metadata", func() {
```
Need to update the test name to reflect what we are really doing here.

pkalever replied:
Looks good, as discussed

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -850,6 +850,220 @@ var _ = Describe(cephfsType, func() {
+				}
+
+				// delete clone and app
+				err = deletePVCAndApp(name, f, pvcClone, appClone)
+				if err != nil {
+					e2elog.Failf("failed to delete PVC and app (%s): %v", f.UniqueName, err)
+				}
```
we dont need app creation here. just create the clone?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -850,6 +850,220 @@ var _ = Describe(cephfsType, func() {
+				}
+				pvc.Namespace = f.UniqueName
+
+				app, err := loadApp(appPath)
+				if err != nil {
+					e2elog.Failf("failed to load app: %v", err)
+				}
```
app creation is not required

Madhu-1 commented on e2e/cephfs_helper.go:
```diff
@@ -212,6 +212,97 @@ func listCephFSSubVolumes(f *framework.Framework, filesystem, groupname string)
+type cephfsSubvolumeMetadata struct {
+	PVCNameKey      string `json:"csi.storage.k8s.io/pvc/name"`
+	PVCNamespaceKey string `json:"csi.storage.k8s.io/pvc/namespace"`
+	PVName          string `json:"csi.storage.k8s.io/pv/name"`
```
PVNameKey?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -850,6 +850,220 @@ var _ = Describe(cephfsType, func() {
+				if len(subvolumes) == 0 {
+					e2elog.Failf("cephFS subvolumes list is empty %s", fileSystemName)
+				}
+				snaps, err := listCephFSSnapshots(f, fileSystemName, subvolumes[0].Name, subvolumegroup)
+				if err != nil {
+					e2elog.Failf("failed to list subvolume snapshots: %v", err)
+				}
+				if len(snaps) == 0 {
+					e2elog.Failf("cephFS snapshots list is empty %s/%s", fileSystemName, subvolumes[0].Name)
+				}
```
This is a good one :+1: . can you please open an issue to add this check for all the tests where we are creating the snapshots. Looks like we are not validating the snapshot counts in existing checks

pkalever replied:
Sure will open one, thanks!

pkalever replied:
@Madhu-1 here is one https://github.com/ceph/ceph-csi/issues/3224

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,82 @@
+	for k, v := range parameters {
+		err := s.SetSnapshotMetadata(k, v)
+		if err != nil {
+			return fmt.Errorf("failed to set metadata key %q, value %q on subvolume snapshot %v: %w", k, v, s.SnapshotID, err)
```
return subvolume/snapshot name for better logging and easy debugging?

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,82 @@
+
+// GetSnapshotMetadata gets custom metadata on the subvolume snapshot in a
+// volume based on key name.
+func (s *snapshotClient) GetSnapshotMetadata(key string) (string, error) {
```
This is not used anywhere right? if not please remove it

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -815,6 +815,18 @@ func (cs *ControllerServer) CreateSnapshot(
 		}
 
+		// Update snapshot-name/snapshot-namespace/snapshotcontent-name details on
+		// sunvolume snapshot as metadata on restart of provisioner pod when
```
no need to mention restart just mention updating metadata in case a snapshot already exists.

Madhu-1 reviewed (changes requested):
some nits. changes LGTM

pkalever commented:
@Madhu-1 addressed all the review comments, please take a look. Many Thanks!

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -815,6 +815,17 @@ func (cs *ControllerServer) CreateSnapshot(
 
+		// Update snapshot-name/snapshot-namespace/snapshotcontent-name details on
+		// subvolume snapshot as metadata in case snapshot already exist
+		parameters := req.GetParameters()
+		if len(parameters) != 0 {
```
this is always not empty as we have at least clusterID. the check should be for len of metadata?

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -860,6 +871,7 @@ func doSnapshot(
 	ctx context.Context,
 	volOpt *store.VolumeOptions,
 	snapshotName string,
+	parameters map[string]string,
```
no need to pass parameters just pass metadata in input

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -895,6 +907,14 @@ func doSnapshot(
+	// Set snapshot-name/snapshot-namespace/snapshotcontent-name details
+	// on subvolume snapshot as metadata on create
+	metadata := k8s.GetSnapshotMetadata(parameters)
+	err = snapClient.SetAllSnapshotMetadata(metadata)
```
also need len(metadata) >0 check here?

Madhu-1 reviewed (commented):
some nits.

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -800,6 +800,16 @@ func (cs *ControllerServer) CreateSnapshot(
 
+		// Update snapshot-name/snapshot-namespace/snapshotcontent-name details on
+		// subvolume snapshot as metadata in case snapshot already exist
+		metadata := k8s.GetSnapshotMetadata(req.GetParameters())
+		if len(metadata) != 0 {
+			err = snapClient.SetAllSnapshotMetadata(metadata)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+		}
```
skip this one if someone is using an older version of ceph which does not support metadata operation?

pkalever replied:
Done now, PTAL.

Madhu-1 reviewed (commented):
changes LGTM. we need to be backward compatible with an older version of ceph.

Commit:
cephfs: handle metadata op-failures with unsupported ceph versions

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,211 @@ var _ = Describe(cephfsType, func() {
+				if pvcCloneObj.Spec.VolumeName == "" {
+					e2elog.Logf("pv clone name is empty %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
+				subvols, err = listCephFSSubVolumes(f, fileSystemName, subvolumegroup)
+				if err != nil {
+					e2elog.Failf("failed to list subvolumes: %v", err)
+				}
+				if len(subvols) == 0 {
+					e2elog.Failf("cephFS subvolumes list is empty %s", fileSystemName)
+				}
+				var cloneSubVol string
+				for _, subvol := range subvols {
+					if subvol.Name == parentSubVol {
+						continue
+					}
+					cloneSubVol = subvol.Name
+
+					break
+				}
```
clone subvolume name already exists on the PV object just get the PV object and use it?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,211 @@ var _ = Describe(cephfsType, func() {
+				if err != nil {
+					e2elog.Failf("failed to create snapshot (%s): %v", snap.Name, err)
+				}
+				subvolumes, err := listCephFSSubVolumes(f, fileSystemName, subvolumegroup)
+				if err != nil {
+					e2elog.Failf("failed to list subvolumes: %v", err)
+				}
+				if len(subvolumes) == 0 {
+					e2elog.Failf("cephFS subvolumes list is empty %s", fileSystemName)
+				}
```
subvolume name already present in the PV object just use it?

pkalever replied:
I hope you mean
```
pvc, err := getPersistentVolumeClaim(c, pvc.Namespace, pvc.Name)
if err != nil {                                                 
    e2elog.Failf("failed to get PVC: %v", err)   
}                                                               
```
And use `pvc.Spec.VolumeName` instead of `subvolumes[0].Name`

Madhu-1 replied:
volumeName is PV name, you need to get PV object and the subvolume name from pv.spec.csi.volumeattributes.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,211 @@ var _ = Describe(cephfsType, func() {
+				if pvcCloneObj.Spec.VolumeName == "" {
+					e2elog.Logf("pv clone name is empty %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
+				subvol, err := listCephFSSubVolumes(f, fileSystemName, subvolumegroup)
+				if err != nil {
+					e2elog.Failf("failed to list subvolumes: %v", err)
+				}
+				if len(subvol) == 0 {
+					e2elog.Failf("cephFS subvolumes list is empty %s", fileSystemName)
+				}
```
same here as well.

Madhu-1 commented on e2e/cephfs_helper.go:
```diff
@@ -212,6 +212,97 @@ func listCephFSSubVolumes(f *framework.Framework, filesystem, groupname string)
+	filesystem,
+	subvolume,
+	groupname string,
+) (cephfsSubvolumeMetadata, error) {
```
```suggestion
) (*cephfsSubvolumeMetadata, error) {
```

Madhu-1 commented on e2e/cephfs_helper.go:
```diff
@@ -212,6 +212,97 @@ func listCephFSSubVolumes(f *framework.Framework, filesystem, groupname string)
+	subvolume,
+	snapname,
+	groupname string,
+) (cephfsSnapshotMetadata, error) {
```
```suggestion
) (*cephfsSnapshotMetadata, error) {
```

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -673,7 +674,7 @@ func (cs *ControllerServer) ControllerExpandVolume(
 // CreateSnapshot creates the snapshot in backend and stores metadata
 // in store
-// nolint:gocyclo,cyclop // golangci-lint did not catch this earlier, needs to get fixed late
+// nolint:gocognit,gocyclo,cyclop // golangci-lint did not catch this earlier, needs to get fixed late
```
:(

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -800,6 +805,16 @@ func (cs *ControllerServer) CreateSnapshot(
 
+		// Update snapshot-name/snapshot-namespace/snapshotcontent-name details on
+		// subvolume snapshot as metadata in case snapshot already exist
+		metadata := k8s.GetSnapshotMetadata(req.GetParameters())
```
get metadata only once and use it? we have two calls for `GetSnapshotMetadata` in createSnapshot

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,114 @@
+
+// SetSnapshotMetadata sets custom metadata on the subvolume snapshot in a
+// volume as a key-value pair.
+func (s *snapshotClient) SetSnapshotMetadata(key, value string) error {
```
we dont need to export this function as its not used outside of this file?

pkalever replied:
I understand the concern. But to be practical, I wanted to align this just like SetMetadata()/GetMetadata/RemoveMetadata(). Let me know if you are not convinced.

Madhu-1 replied:
we(at least I) might have missed that one. if the function is not used outside of the package, it should not be exported, we can revisit the existing functions and correct them.

pkalever replied:
Sure I can fix it.
I thought we were writing those functions assuming there will be some direct consumers in the future calling them from outside this file.

Madhu-1 replied:
Yes but IMHO as there are no real users we shouldn't expose it, expose it when/if required always

Madhu-1 commented on internal/cephfs/core/snapshot_metadata.go:
```diff
@@ -0,0 +1,114 @@
+
+// RemoveSnapshotMetadata removes custom metadata set on the subvolume
+// snapshot in a volume using the metadata key.
+func (s *snapshotClient) RemoveSnapshotMetadata(key string) error {
```
same here as well

Madhu-1 reviewed (commented):
some nits.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,174 @@ var _ = Describe(cephfsType, func() {
+				if err != nil {
+					e2elog.Logf("error getting pvc %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
+				if pvcCloneObj.Spec.VolumeName == "" {
+					e2elog.Logf("pv clone name is empty %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
```
why this check is required?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,174 @@ var _ = Describe(cephfsType, func() {
+				err = deleteResource(cephFSExamplePath + "snapshotclass.yaml")
+				if err != nil {
+					e2elog.Failf("failed to delete CephFS snapshotclass: %v", err)
+				}
```
add check to validate subvolumes count at end of the test?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,174 @@ var _ = Describe(cephfsType, func() {
+				if err != nil {
+					e2elog.Logf("error getting pvc %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
+				if pvcCloneObj.Spec.VolumeName == "" {
+					e2elog.Logf("pv clone name is empty %q in namespace %q: %v", pvcClone.Name, pvcClone.Namespace, err)
+				}
```
not required?

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,174 @@ var _ = Describe(cephfsType, func() {
+				err = deletePVCAndValidatePV(f.ClientSet, pvcClone, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to delete pvc clone: %v", err)
+				}
```
add check to validate subvolumes count at end of the test?

Madhu-1 reviewed (dismissed):
small nits in e2e. changes LGTM

pkalever commented on e2e/cephfs.go:
```diff
@@ -848,6 +848,172 @@ var _ = Describe(cephfsType, func() {
+					e2elog.Failf("failed to delete PVC: %v", err)
+				}
+
+				// Test Restore snapshot
```
@Madhu-1 what kind of metadata details do we expect on the PVC which is restored from a snapshot?
I'm I missing anything below, I see the control is entering condition `if cloneMetadata.PVCNameKey != pvcClone.Name {}` all the time.

Madhu-1 replied:
check we have the metadata support in 17.2.1, which is the default in our CI

pkalever replied:
Yes, we do! I'm rebasing this now, lets wait and see as you previously suggested.

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.24

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22
